ffi_internals/src/native_type_data.rs

```diff
@@ -105,6 +114,154 @@ impl From<(NativeType, WrappingType)> for NativeTypeData {
+                vec: true,
+                result: unparsed.result,
+            }),
+            Type::Path(ty) => {
+                let segment = ty.path.segments.last().unwrap();
+                let ident = segment.ident.clone();
+                if ident == format_ident!("Option") {
+                    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+                        if let syn::GenericArgument::Type(arg) = arguments.args.first().unwrap() {
+                            Self::from(UnparsedNativeTypeData {
+                                ty: arg.clone(),
+                                option: true,
+                                vec: unparsed.vec,
+                                result: unparsed.result,
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Option: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Option: {:?}", segment);
+                    }
+                } else if ident == format_ident!("Result") {
+                    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+                        // Take the first generic argument, which is the success type.
+                        if let syn::GenericArgument::Type(arg) = arguments.args.first().unwrap() {
+                            Self::from(UnparsedNativeTypeData {
+                                ty: arg.clone(),
+                                option: unparsed.option,
+                                vec: unparsed.vec,
+                                result: true,
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Result: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Result: {:?}", segment);
+                    }
+                } else if ident == format_ident!("Vec") {
+                    if let syn::PathArguments::AngleBracketed(arguments) = &segment.arguments {
+                        if let syn::GenericArgument::Type(arg) = arguments.args.first().unwrap() {
+                            Self::from(UnparsedNativeTypeData {
+                                ty: arg.clone(),
+                                option: unparsed.option,
+                                vec: true,
+                                result: unparsed.result,
+                            })
+                        } else {
+                            panic!("Unexpected arguments for Vec: {:?}", arguments);
+                        }
+                    } else {
+                        panic!("Unexpected segment contents for Vec: {:?}", segment);
+                    }
+                } else {
+                    let native_type = NativeType::from(ident);
```

Comment:
Is there a nice path for breaking some of this up? Looks like similar shapes in each of those logic gates...